homeassistant/components/arcam_fmj/media_player.py

```diff
@@ -186,6 +145,57 @@ async def async_update(self):
         """Force update of state."""
         _LOGGER.debug("Update state %s", self.name)
         await self._state.update()
+        self._attr_state = STATE_ON if self._state.get_power() else STATE_OFF
+        self._attr_device_info = {
+            "name": self._device_name,
+            "identifiers": {
+                (DOMAIN, self._uuid),
+                (DOMAIN, self._state.client.host, self._state.client.port),
+            },
+            "model": "Arcam FMJ AVR",
+            "manufacturer": "Arcam",
+        }
+        source = self._state.get_source()
+        self._attr_source = source.name if source else None
+        self._attr_source_list = [x.name for x in self._state.get_source_list()]
+        if self._state.zn == 1:
+            if self._get_2ch():
+                value = self._state.get_decode_mode_2ch()
+                self._attr_sound_mode_list = [x.name for x in DecodeMode2CH]
+            else:
+                value = self._state.get_decode_mode_mch()
+                self._attr_sound_mode_list = [x.name for x in DecodeModeMCH]
+            self._attr_sound_mode = value.name if value else None
+        else:
+            self._attr_sound_mode = self._attr_sound_mode_list = None
+        value = self._state.get_mute()
+        self._attr_is_volume_muted = value if value else None
+        value = self._state.get_volume()
+        self._attr_volume_level = value / 99.0 if value else None
+        if source == SourceCodes.DAB:
+            self._attr_media_content_type = MEDIA_TYPE_MUSIC
+            self._attr_media_channel = self._state.get_dab_station()
+            self._attr_media_artist = self._state.get_dls_pdt()
+        elif source == SourceCodes.FM:
+            self._attr_media_content_type = MEDIA_TYPE_MUSIC
+            self._attr_media_channel = self._state.get_rds_information()
+        else:
+            self._attr_media_content_type = (
+                self._attr_media_channel
+            ) = self._attr_media_artist = None
+        if source in (SourceCodes.DAB, SourceCodes.FM):
+            preset = self._state.get_tuner_preset()
+            self._attr_media_content_id = f"preset:{preset}" if preset else None
+        else:
+            self._attr_media_content_id = None
+        if source:
+            channel = self._attr_media_channel
+            if channel:
+                self._attr_media_title = f"{source.name} - {channel}"
+            else:
+                self._attr_media_title = source.name
+        else:
+            self._attr_media_title = None
```

Comment:
This function ends up too large and complex now. I still want several of these things to be sub functions.

Reply:
Undone.

Reply:
How would you like to see this done?

Reply:
I'm unsure exactly how. But keeping each calculated property as a function and just calling them from update method to set the attribute members. 

The benefits are somewhat limited :)

Reply:
Perhaps the media attributes can be set in a separate method called by `async_update`?

Reply:
Yes that might be enough. async_update_media()